### ISSUE TITLE
Pseudo dice while training and validation

### ISSUE THREAD

**Issue by Body123** [MIC-DKFZ/nnUNet#2128]
I want to ask while training in the logs I found as an example :
2024-04-23 15:22:40.711169: Epoch 195
2024-04-23 15:22:40.724555: Current learning rate: 0.00036
2024-04-23 15:23:41.803760: train_loss -0.7526
2024-04-23 15:23:41.832845: val_loss -0.771
2024-04-23 15:23:41.869548: Pseudo dice [0.8399]
2024-04-23 15:23:41.895059: Epoch time: 61.1 s
2024-04-23 15:23:41.911069: Yayy! New best EMA pseudo Dice: 0.8196
but when calculating after the number of epochs end in this fold the number of pseudo dice is less than the above :
2024-04-23 15:29:06.880623: Mean Validation Dice: 0.6980713938291729
so I want to ask what is this number represents 2024-04-23 15:23:41.869548: Pseudo dice [0.8399] and what this represent :
2024-04-23 15:23:41.911069: Yayy! New best EMA pseudo Dice: 0.8196

**Comment by Chasel-Chen** [comment]
Here is my understanding. The nnUNet is trained based on patches, so the actual input data fed into the network are patches sampled from the original image. The dice metric during training is also calculated based on these patches. After training, nnUNet performs sliding window inference by traversing the image. The patches obtained from this sliding window sampling may differ from the sampling strategy used during training, resulting in a potential decrease in the Dice score.

**Comment by FabianIsensee** [comment]
Exactly as @Chasel-Chen explains it!
The pseudo dice is just to see the training progress and to determine whether overfitting is a problem (green line goes down = bad!). It is not directly related to the actual validation dice! How strongly these two agree depends on the dataset

**Comment by Chasel-Chen** [comment]
Hi，Fabian
I also have a similar confusion. In certain segmentation tasks, patch-based metrics perform exceptionally well during training. However, when performing inference on the original images, there is a significant difference in Dice scores, even when evaluated on the training set. I have attempted to modify the sampling strategy (such as shifting the generated bounding boxes), but without much success. Are there any suggestions for further experimentation and improvement, such as adjusting patch size, oversample rate, or any other methods? 
Thank you very much.

**Comment by FabianIsensee** [comment]
There is no way of 'fixing' this, please just take it as it is! It is not supposed to be the same as the validation dice. Running actual validations would take too long

**Comment by Body123** [comment]
but what is the difference between these:
2024-04-23 15:23:41.869548: Pseudo dice [0.8399]
2024-04-23 15:23:41.911069: Yayy! New best EMA pseudo Dice: 0.8196
each one of them is related to training and which one related to validation

**Comment by FabianIsensee** [comment]
EMA stands for exponential moving average. This tracks the pseudo dice following a `updated = old * alpha + (1 - alpha) * new` formula. This is the smoothed green line in the progress.png plot

**Comment by Body123** [comment]
Okay thanks alot, my last question, and sorry for many questions: those numbers are based on taking some random patch from the training set during epoch not the whole image or they are from validation set during the epoch  :
2024-04-23 15:23:41.869548: Pseudo dice [0.8399]
2024-04-23 15:23:41.911069: Yayy! New best EMA pseudo Dice: 0.8196

**Comment by FabianIsensee** [comment]
they are random patches from the validation set, sampled according to the same rules as the training patches. Dice is computed over all sampled patches (pretending all patches together are one image) and not for each patch

**Comment by mehnaz1985** [comment]
![image](https://github.com/MIC-DKFZ/nnUNet/assets/66414863/550130b2-a7f2-4f71-8cce-cbc9d7cdd593)
 Can anyone tell me what is wrong with my training? Any help is much appreciated. @Body123 @Chasel-Chen @FabianIsensee

**Comment by FabianIsensee** [comment]
Seems like
1) your epoch time is really slow, see https://github.com/MIC-DKFZ/nnUNet/blob/master/documentation/benchmarking.md
2) you seem to suffer from some overfitting
3) one class is not learned properly

What could concretely be the problem I cant say just from what you shared

**Comment by mehnaz1985** [comment]
Thank you for your response, @FabianIsensee . Let me explain my task and the dataset. I am attempting to segment recurrent and non-recurrent areas in the peritumoral region. There are three labels, including background, and I have seven channels (t1, t2, t1c, flair, dti_fa, dti_ra, and dti_md). FYI, the recurrent area is tiny compared to the non-recurrent area, possibly a class imbalance problem. I was thinking of giving more weight to the minority class in the loss function, but I was confused about where to change. There are several losses! and also need to call that loss in the nnUNetTrainer.py? pls enlighten me in this matter.

**Comment by mehnaz1985** [comment]
now another problem arose, 
![image](https://github.com/MIC-DKFZ/nnUNet/assets/66414863/6e2f2b41-e756-4d90-82fd-ab025afc6798)

**Comment by FabianIsensee** [comment]
It looks like you created a class that has the same name as the nnUNetTrainer but does not inherit from it? When making changes, please remember to create trainer classes with unique names and make them inherit from nnUNetTrainer.

It seems like one of your classes is very difficult to learn or was annotated too inconsistently. Giving it more weight will not solve the problem. Looking at your training loss it seems like the model can learn the class on the training data but it fails to generalize to the validation set

**Comment by dancingreindeer** [comment]
> It looks like you created a class that has the same name as the nnUNetTrainer but does not inherit from it? When making changes, please remember to create trainer classes with unique names and make them inherit from nnUNetTrainer.
> 
> It seems like one of your classes is very difficult to learn or was annotated too inconsistently. Giving it more weight will not solve the problem. Looking at your training loss it seems like the model can learn the class on the training data but it fails to generalize to the validation set

Hi Fabian, I would like to continue this discussion a bit. 

1. How do we determine if annotations are too inconsistent? This actually happens a lot in medical imaging. There are often times when annotations have high inter- and even intra- variability. Are these labels just impossible to train?
2. Besides improving annotation quality (which is often impossible or would take tons of time), is there any other way to make model learns more on the rare labels?

Here are my two cents:
Answer to question 1: I think when you see overfitting on that specific label (good performance on training but bad performance on validation), that usually indicates the annotation is inconsistent. But I am not sure what experiments could be done on the labeling to prove this -- as for most of the time, MLE in medical imaging are not experts in medical knowledge or pathology themselves.

Answer to question 2: I have tried assigning heavier weights to the less learned class, not helping much.

**Comment by MenghanLiu212** [comment]
Hi,

I got nan dice when running training and get validation as in the picture, any help would be highly appreciated!
@FabianIsensee 
Thanks!
<img width="938" alt="Screenshot 2024-11-06 at 4 05 04 AM" src="https://github.com/user-attachments/assets/145acf78-51a0-46e6-a583-90a455b762d8">

**Comment by adamszummer-digica** [comment]
> Hi,
> 
> I got nan dice when running training and get validation as in the picture, any help would be highly appreciated! @FabianIsensee Thanks! <img alt="Screenshot 2024-11-06 at 4 05 04 AM" width="938" src="https://private-user-images.githubusercontent.com/57054930/383546208-145acf78-51a0-46e6-a583-90a455b762d8.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzUxNDU2OTksIm5iZiI6MTczNTE0NTM5OSwicGF0aCI6Ii81NzA1NDkzMC8zODM1NDYyMDgtMTQ1YWNmNzgtNTFhMC00NmU2LWE1ODMtOTBhNDU1Yjc2MmQ4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMjUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjI1VDE2NDk1OVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTU0ODgxOGQxN2Y0ZjlkMTQ4YWMyOWMzYzIyYzMzNWU2YjQ0YWJhZjVhMTljMjg3YzY2Y2Q3NDE5MzVmZjI3ZjgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.7oDQTuYivS5xxD19wCjdIQEl_OuAkh7WQWCINfi1lao">

I have the same my guess is that probably you get NaN when calculating the dice so when you calculate mean you get a NaN as well. I have some classes that are not represented in the dataset, or Dice is unable to be calculated as well.

![image](https://github.com/user-attachments/assets/2f49ed2a-539e-4a6a-b673-ca67be5176c1)
